Make trace interface uniform between module exports and plugins

kjin opened the pull request:
This change:

- Makes `PluginAPI` a per-plugin object that can accept a plugin name
- Overlays a "public API" object over `PluginAPI` that can switch between it and a no-op implementation

matthewloring commented on index.js:
```diff
@@ -83,160 +63,115 @@ var initConfig = function(projectConfig) {
   return config;
 };
 
+var agent = pluginInterface.create('custom');
```
Maybe this could be more descriptive? `'Custom Span API'`?

kjin replied:
That sounds like a good name, I'll change it to that

matthewloring commented on test/test-agent-stopped.js:
```diff
@@ -21,9 +21,13 @@ var http = require('http');
 describe('test-agent-stopped', function() {
   var agent;
-  before(function() {
+  before(function(done) {
```
These changes will be obsolete after https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/pull/410

kjin replied:
OK, will address during rebase

matthewloring commented on src/trace-plugin-interface.js:
```diff
@@ -219,9 +220,80 @@ PluginAPI.prototype.constants = constants;
+        return api.isActive();
+      },
+      initialize_: function(agent) {
+        api = new PluginAPI(agent, pluginName);
```
Do we want to add an assert to ensure this isn't called multiple times?

kjin replied:
I added error logging, though through the publicly exposed interface, neither of these functions will be called multiple times.

matthewloring commented on src/trace-plugin-interface.js:
```diff
@@ -219,9 +220,80 @@ PluginAPI.prototype.constants = constants;
+      initialize_: function(agent) {
+        api = new PluginAPI(agent, pluginName);
+      },
+      disable_: function() {
```
And here?

kjin replied:
Same as ^

matthewloring commented on src/trace-plugin-interface.js:
```diff
@@ -111,9 +111,10 @@ var nullSpan = {
+function PluginAPI(agent, pluginName) {
   this.agent_ = agent;
   this.logger_ = agent.logger;
+  this.pluginName_ = pluginName;
```
Is this used?

kjin replied:
It isn't, but I wanted to include it anyway to make the follow-up change that adds it to log messages more straightforward. In fact, I think most of that is small enough that I could do it in this PR.

matthewloring replied:
Up to you. If you include the logic here, make sure the test changes from https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/pull/392/ are also reflected.

kjin replied:
I'll just do it now, then.

matthewloring commented:
It would also be good to add a few tests to enforce the lifecycle of the agent (cannot be re-initialized after disabling).

kjin commented:
Since `fix for v0.12`:

* Rebased on master
* Addressed comments
* Added plugin name to logs emitted from within `trace-interface.js`
* Moved tests related to the custom span API to `test-trace-interface.js` (they also apply to the plugin interface as well)
* Added new tests for the custom span API to `test-trace-interface.js` as well
* Added fix for a test that was also addressed in #392

@matthewloring I've decided to keep the facade to two states without enforcing anything about state changing, since throwing an error complicates forcing a new agent between tests.

matthewloring commented on src/trace-api.js:
```diff
@@ -174,15 +173,15 @@ PluginAPI.prototype.runInRootSpan = function(options, fn) {
     return new ChildSpan(this.agent_, childContext);
   } else {
-    this.logger_.warn(options.name + ': Attempted to create child span ' +
+    this.logger_.debug(this.pluginName_ + ': Attempted to create child span ' +
```
Could we add a `TODO` on this to fix once we can disambiguate context loss from sampling?

kjin replied:
Done

matthewloring commented on src/trace-api.js:
```diff
@@ -208,20 +208,91 @@ PluginAPI.prototype.wrap = function(fn) {
+ */
+module.exports = function TraceApi(pluginName) {
+  var impl = phantomApiImpl;
+  Object.assign(this, {
```
`Object.assign` is not available in 0.12.

kjin replied:
Forgot about this -- rebasing now to use new `extend` depedency.

matthewloring commented on src/trace-api.js:
```diff
@@ -194,9 +193,10 @@ PluginAPI.prototype.createChildSpan = function(options) {
+TraceApiImplementation.prototype.wrap = function(fn) {
   if (!this.agent_.namespace) {
-    this.logger_.warn('Trace agent: No CLS namespace to bind function to');
+    this.logger_.warn(this.pluginName_ + ': No CLS namespace to bind ' +
```
I know this is a copy but could we fix the grammar. Maybe "No CLS namespace available to bind function"?

kjin replied:
Done

matthewloring commented on src/trace-plugin-loader.js:
```diff
@@ -57,14 +57,18 @@ function activate(agent) {
+    var api = new TraceApi(moduleName);
+    api.enable_(agent);
     plugins[moduleName] = {
+      name: moduleName,
```
Where is this field used?

kjin replied:
Oops, it's an artifact of an earlier revision. Removing this.

matthewloring commented on test/test-grpc-context.js:
```diff
@@ -60,7 +61,7 @@ Object.keys(versions).forEach(function(version) {
       common.replaceDebugLogger(agent, function(error, uri) {
         if (error.indexOf('http') !== -1) {
-          assert.notStrictEqual(uri.indexOf('localhost'), -1);
+          httpDebugPrinted = true;
```
This doesn't check the same property as before. Why was this change needed?

kjin replied:
The test was broken before (in a similar manner to `test-plugins-sample-warning.js`) since the condition `error.indexOf('http') !== -1` was never true. The old assert checks the format of the message being printed in the `http` hook; the format has since changed. Asserting in `afterEach` rather than here just makes sure we don't (erroneously) fail if additional debug-level messages are printed.

matthewloring commented on src/trace-api.js:
```diff
@@ -207,20 +212,91 @@ PluginAPI.prototype.wrap = function(fn) {
 
-PluginAPI.nullSpan = nullSpan;
+TraceApiImplementation.prototype.isActive = function() {
+  return true;
```
Could we replace this with a check for the phantom api in the public interface?

kjin replied:
Done

matthewloring commented on test/test-trace-api.js:
```diff
@@ -0,0 +1,183 @@
+var common = require('./plugins/common.js');
+var TracingPolicy = require('../src/tracing-policy.js');
+
+var config = Object.assign({}, require('../config.js').trace,
```
Also unsafe on 0.12.

kjin replied:
Done

kjin commented:
TBR: @ofrobots

ofrobots commented on src/trace-api.js:
```diff
@@ -207,20 +212,86 @@ PluginAPI.prototype.wrap = function(fn) {
 
-module.exports = PluginAPI;
+/**
+ * Creates an object that provides an interface to the trace agent
```
nit: fix word wrapping for this comment.

kjin replied:
That puts it over 80 characters... I'll reword it though.

ofrobots commented on test/plugins/common.js:
```diff
@@ -235,7 +236,11 @@ function avoidTraceWriterAuth(agent) {
 
 function stopAgent(agent) {
-  agent.private_().stop();
+  if (agent.isActive()) {
```
Ideally, once the comments I expressed in `index.js` are addressed, this function be removed.

ofrobots commented on index.js:
```diff
@@ -215,16 +66,100 @@ var publicAgent = {
+    return traceApi;
+  }
+
+  agent = require('./src/trace-agent.js').get(config, logger);
```
Any reason why this require is not done at the top?

Secondly, we have three variables here that have a strong dependence on each other. `agent`, `traceApi`, and `pluginLoader`. 

* First of all can we move `pluginLoader` into the agent? it should be a concept owned by the agent. Since there is a strong dependence on `pluginLoader.activate` and `deactivate`, when `agent.start` and `stop` are called respectively, it would better if `agent` was *responsible* for calling these? I see that you moved this from the agent to here. Is there a reason for doing this?
* Similarly, can we make the `agent` responsible for calling `traceApi.enable` and `disable` from `start` and `stop` respectively?
* Nit: we have three names for similar operations: `{start, enable, activate}` and `{stop, disable, deactivate}`. Do we really need all three?

kjin replied:
I agree that the function calls that occur when one starts or stops the agent should be coalesced in one place, and I think it should be in `index.js` because my thoughts were that `TraceApiImplementation` and `TraceAgent` would eventually be folded into a single object (because one is really just a thin layer over the other, and this extra layer is unnecessary). I pulled `pluginLoader` out to take a step in that direction, because `TraceApiImplementation` in my opinion should not be responsible for it.

For the sake of readability I'd be OK with moving it back (and also reducing the number of synonyms on `start` and `stop`). But hopefully that gives a better understanding of the rationale behind this move.
